driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java

```diff
@@ -286,7 +289,11 @@ public void refreshSchema(String keyspace, String table) throws InterruptedExcep
             // At startup, when we add the initial nodes, this will be null, which is ok
             if (c == null)
                 return;
-            refreshSchema(c, keyspace, table, cluster, false);
+            refreshSchema(c, keyspace, table, cluster);
+            // If the table is null, we either rebuild all from scratch or have an updated keyspace. In both cases, rebuild the token map
+            // since some replication on some keyspace may have changed
+            if (table == null)
```

Comment:
Same remark as above, but here the node list refresh can be done asynchronously.